### PR TITLE
Properly handle responses that should trigger download action

### PR DESCRIPTION
**Pull request by miasma13**
Task/Issue URL: https://app.asana.com/0/414709148257752/1208359719493827/f

**Description**:
We should be more deliberate on handling download triggers e.g. with links with `download` attribute, BLOB content, responses with `Content-Disposition: attachment` header. 

**Steps to test this PR**:
See parent task for set of download scenarios to test. _Please compare behavior with version from before changes._

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -816,7 +817,6 @@
 		B6BA95C328891E33004ABA20 /* BrowsingMenuAnimator.swift in Sources */ = {isa = PBXBuildFile; fileRef = B6BA95C228891E33004ABA20 /* BrowsingMenuAnimator.swift */; };
 		B6BA95C528894A28004ABA20 /* BrowsingMenuViewController.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = B6BA95C428894A28004ABA20 /* BrowsingMenuViewController.storyboard */; };
 		B6BA95E828924730004ABA20 /* JSAlertController.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = B6BA95E728924730004ABA20 /* JSAlertController.storyboard */; };
-		B6CB93E5286445AB0090FEB4 /* Base64DownloadSession.swift in Sources */ = {isa = PBXBuildFile; fileRef = B6CB93E4286445AB0090FEB4 /* Base64DownloadSession.swift */; };
```
Comment (miasma13) [inline DuckDuckGo.xcodeproj/project.pbxproj:819]: No longer used anywhere.

File: Core/PixelEvent.swift [hunk 2]
```diff
@@ -1326,7 +1325,6 @@ extension Pixel.Event {
         case .cachedTabPreviewRemovalError: return "m_d_tpre"
             
         case .missingDownloadedFile: return "m_d_missing_downloaded_file"
-        case .unhandledDownload: return "m_d_unhandled_download"
```
Comment (miasma13) [inline Core/PixelEvent.swift:1329]: Looking at the code path and confirming in Kibana this pixel is/was not being fired.

File: DuckDuckGo.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -6517,6 +6516,7 @@
 				F1D43AF92B99C1D300BAB743 /* BareBonesBrowserKit */,
 				9F8FE9482BAE50E50071E372 /* Lottie */,
 				9F96F73A2C9144D5009E45D5 /* Onboarding */,
+				1E5918462CA422A7008ED2B3 /* Navigation */,
```
Comment (miasma13) [inline DuckDuckGo.xcodeproj/project.pbxproj:6519]: Required for `WKNavigationAction`, `HTTPURLResponse` and other navigation related extensions.

File: DuckDuckGo/TabViewController.swift [hunk 4]
```diff
@@ -2140,31 +2149,38 @@ extension TabViewController {
...
-                    }
-                    callback(self.transfer(download,
-                                           to: downloadManager,
-                                           with: navigationResponse.response,
-                                           suggestedFilename: suggestedFilename,
-                                           isTemporary: isTemporary))
+                let hasNavigationActionRequestedDownload = (self.recentNavigationActionShouldPerformDownloadURL != nil) && self.recentNavigationActionShouldPerformDownloadURL == navigationResponse.response.url
+                let canLoadOrPreviewTheFile = navigationResponse.canShowMIMEType || FilePreviewHelper.canAutoPreviewMIMEType(downloadMetadata.mimeType)
 
-                } else {
```
Comment (miasma13) [inline DuckDuckGo/TabViewController.swift:2156]: Logical order of if-else has been reversed here.

File: DuckDuckGo/TabViewController.swift [hunk 5]
```diff
@@ -167,6 +168,9 @@ class TabViewController: UIViewController {
     var mostRecentAutoPreviewDownloadID: UUID?
     private var blobDownloadTargetFrame: WKFrameInfo?
 
+    // Recent request's URL if its WKNavigationAction had shouldPerformDownload set to true
+    private var recentNavigationActionShouldPerformDownloadURL: URL?
```
Comment (miasma13) [inline DuckDuckGo/TabViewController.swift:172]: This is used to hold and pass information from `WKNavigationAction`'s  `shouldPerformDownload` read in: 
- `func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void)`

to be used for decision logic later in:
- `func webView(_ webView: WKWebView, decidePolicyFor navigationResponse: WKNavigationResponse, decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void)`

and
- `func webView(_ webView: WKWebView, navigationResponse: WKNavigationResponse, didBecome download: WKDownload)`

File: DuckDuckGo/TabViewController.swift [hunk 6]
```diff
@@ -1742,6 +1758,8 @@ extension TabViewController: WKNavigationDelegate {
         }
 
         if navigationAction.isTargetingMainFrame(),
+           !navigationAction.isSameDocumentNavigation,
+           !navigationAction.shouldDownload,
```
Comment (miasma13) [inline DuckDuckGo/TabViewController.swift:1767]: These are required changes for GPC protected sites not to break IFrame triggered downloads, as canceling and re-triggering request breaks the download attribute. This matches the logic on macOS where for page sub-requests we do not trigger the full distributed navigation chain execution. Aligned and accepted by the privacy team.

File: DuckDuckGo/TabViewController.swift [hunk 7]
```diff
@@ -1317,48 +1322,57 @@ extension TabViewController: WKNavigationDelegate {
...
 
-        if navigationResponse.canShowMIMEType && !FilePreviewHelper.canAutoPreviewMIMEType(mimeType) {
+        // HTTP response has "Content-Disposition: attachment" header
+        let hasContentDispositionAttachment = httpResponse?.shouldDownload ?? false
+        // If preceding WKNavigationAction requested to start the download
+        let hasNavigationActionRequestedDownload = (recentNavigationActionShouldPerformDownloadURL != nil) && recentNavigationActionShouldPerformDownloadURL == navigationResponse.response.url
+        // File can be rendered by web view or in custom preview handled by FilePreviewHelper
+        let canLoadOrPreviewTheFile = navigationResponse.canShowMIMEType || FilePreviewHelper.canAutoPreviewMIMEType(mimeType)
+
+        let shouldTriggerDownloadAction = hasContentDispositionAttachment || hasNavigationActionRequestedDownload || !canLoadOrPreviewTheFile
```
Comment (brindy) [inline DuckDuckGo/TabViewController.swift]: None of this is used more than once and the `shouldTriggerDownloadAction` is used *after* a bunch of other evaluations.  How about making that a private function that is called instead at the time it's relevant?

File: DuckDuckGo/TabViewController.swift [hunk 8]
```diff
@@ -1317,48 +1322,57 @@ extension TabViewController: WKNavigationDelegate {
...
+                decisionHandler(.download)
+            }
+        } else if FilePreviewHelper.canAutoPreviewMIMEType(mimeType) {
+            // 2. For this MIME type we are able to provide a better custom preview via FilePreviewHelper so it takes priority
+            let download = self.startDownload(with: navigationResponse, decisionHandler: decisionHandler)
+            mostRecentAutoPreviewDownloadID = download?.id
+            Pixel.fire(pixel: .downloadStarted,
+                       withAdditionalParameters: [PixelParameters.canAutoPreviewMIMEType: "1"])
+        } else if shouldTriggerDownloadAction,
+                  let downloadMetadata = AppDependencyProvider.shared.downloadManager.downloadMetaData(for: navigationResponse.response) {
```
Comment (brindy) [inline DuckDuckGo/TabViewController.swift:1343]: Now would be a really nice time to refactor out this singleton 😇

**Review by brindy — CHANGES_REQUESTED** [review]
~~Something isn't quite right.  I happened to be on this page and it triggers a download~~

~~https://duckduckgo.com/email/~~

~~But also doing a search triggers the download.~~

Edit: my fault, I changed some code when testing.

**Review by brindy — APPROVED** [review]
LGTM, thanks!

**Pull request merged**